hera_pspec/grouping.py

```diff
@@ -1712,3 +1712,263 @@ def list_of_lists_of_tuples(s):
+    extra delay will be ignored.
+    
+    Parameters
+    ----------
+    uvp
+        The UVPSpec object to average.
+    kernel
+        The averaging kernel. This should be a 1D array containing the averaging weights
+        within each non-overlapping bin. It will be normalized to have a sum of unity.
+        The size of the kernel will define the size of the resulting delay bins.
+    cov_weighting_for_pn
+        Whether to average the P_N statistic using known covariance information, or 
+        simply by averaging with the kernel.
+    zero_bin_kernel
+        The kernel to use for the zero bin. This must be symmetric and have an odd
+        length. By default, this is a kernel of length 1 with value 1.
+        This means that the zero bin is not averaged at all. 
+        
+    Returns
+    -------
+    uvp
+        A new uvp object with the down-sampled delays.
+    """
+    uvp  = copy.deepcopy(uvp)
+    stats_array = getattr(uvp, 'stats_array', {})
+    if hasattr(uvp, "cov_array_real"):
+        cov_weighted_stats = tuple(stat for stat in cov_weighted_stats if stat in stats_array)
+    else:
+        cov_weighted_stats = ()    
+
+    kernel = np.asarray(kernel, dtype=float)
+    if not kernel.ndim == 1:
+        raise ValueError("The kernel must be 1D!")
+
+    nk = len(kernel)
+    if nk > len(uvp.dly_array) // 2:
+        raise ValueError("The kernel size must be smaller than half the size of the delay array!")
+    
+    zero_bin_kernel = np.asarray(zero_bin_kernel, dtype=float)
+    if not np.allclose(zero_bin_kernel, zero_bin_kernel[::-1]):
+        raise ValueError("The zero bin kernel must be symmetric!")
+    
+    kernel /= np.sum(kernel)
+    zero_bin_kernel /= np.sum(zero_bin_kernel)
+    
+    # We will replace parts of the new_uvp with downsampled
+    # delays.
+    new_uvp = copy.deepcopy(uvp)
+
+    new_uvp.data_array = {}
+    new_uvp.stats_array = {name: {} for name in stats_array}
+
+    dly_array = []
+    spw_dly_array = []
+    for spw in uvp.spw_array:   
+        
+        p = uvp.data_array[spw]
+        dly = uvp.get_dlys(spw)
+        slices, kernels = _get_delay_slices(dly, kernel=kernel, zero_kernel=zero_bin_kernel)
+        newn = len(slices)
+
+        new_uvp.data_array[spw] = _bin_data_like_array(p, kernels, slices)
+        newdly = np.array([np.sum(dly[slc]*krn) for slc, krn in zip(slices, kernels)])
+        dly_array.append(newdly)
+
+        stats = {
+            name: np.sqrt(_bin_data_like_array(
+                stats_array[name][spw]**2, kernels, slices
+            ))
+            for name in stats_array
+            if name not in cov_weighted_stats
+        }
+        if hasattr(uvp, "cov_array_real"):
+            oldcov = uvp.cov_array_real[spw]
+            newcov = _bin_cov_like_array(oldcov, kernels, slices)
+            new_uvp.cov_array_real[spw] = newcov
+            new_uvp.cov_array_imag[spw] = np.zeros_like(new_uvp.cov_array_real[spw])
```

Comment:
Shouldn't you bin the imaginary part of the cov matrix as well? (it looks like currently it is always an array of zeros if it comes from `UVPSpec` but just in case this changes in the future...)